examples/chef/common/clusters/media-playback/MediaPlaybackManager.cpp

```diff
@@ -15,22 +15,41 @@
  *    limitations under the License.
  */
 
-#include <app/util/config.h>
-#ifdef MATTER_DM_PLUGIN_MEDIA_PLAYBACK_SERVER
-#include "MediaPlaybackManager.h"
 #include <app-common/zap-generated/attributes/Accessors.h>
 #include <app/util/config.h>
-
+#include <map>
 #include <string>
+#ifdef MATTER_DM_PLUGIN_MEDIA_PLAYBACK_SERVER
+#include "MediaPlaybackManager.h"
 
 using namespace std;
+using namespace chip;
+using namespace chip::app;
 using namespace chip::app::DataModel;
 using namespace chip::app::Clusters::MediaPlayback;
-using namespace chip::Uint8;
 
 using chip::CharSpan;
 using chip::app::AttributeValueEncoder;
 using chip::app::CommandResponseHelper;
+using chip::Protocols::InteractionModel::Status;
+
+MediaPlaybackManager::MediaPlaybackManager(chip::EndpointId endpoint) : mEndpoint(endpoint)
+{
+    // Sync the attributes from attribute storage
+    Status status = Attributes::CurrentState::Get(endpoint, &mCurrentState);
```

Comment:
This is really quite odd.  Why are we storing the data in two places and keeping them in sync?  Why is the responsibility for doing that (or for triggering reporting in general) in the _application_ and not in the _cluster implementation_?

Ideally the way this would work, for simple data like this, is:

1. mCurrentState would live in the cluster implementation.
2. When the state needs to change, a setter on the cluster implementation would be called to set the state.
3. The cluster would then mark things dirty as needed.

Reply:
> Why are we storing the data in two places and keeping them in sync?
The `mCurrentState` is a variable in the RAM and could not be persistently stored, right? Considering if this is an application, you changed the variables, killed the process, and run the program again. You just reload the value from the storage instead of reset it, right?

> 1. mCurrentState would live in the cluster implementation.
I don't quite understand this. `mCurrentState` still live in the cluster. What I'm doing is just restoring the value from anywhere it stored in last change. The current example is restoring it from attribute storage.

> 2. When the state needs to change, a setter on the cluster implementation would be called to set the state.
> 3. The cluster would then mark things dirty as needed.
The cluster would then mark things dirty as needed.
 I assume the new function `MediaPlaybackManager::HandleSetCurrentState(...)` already fulfill this?

Reply:
Let's remove mCurrentState and completely rely on the AttributeStore view if we add `Attributes::CurrentState::Get`/`Set` for attributes.

Reply:
> Let's remove mCurrentState and completely rely on the AttributeStore view if we add Attributes::CurrentState::Get/Set for attributes.

This seems making sense. My understanding is the Attribute access logics are delegated to the application, so the application could decide how to read/write/access the values. Reducing the variable seems can reduce potential risks of out-of-sync.

Reply:
> The mCurrentState is a variable in the RAM and could not be persistently stored, right?

So is what you do with Attributes::Whatever::Set, by default, unless it's either marked as NVM in the ZAP config or marked external and the external attribute handler does persistence.

> My understanding is the Attribute access logics are delegated to the application

They might or might not be, depending on the ZAP configuration.

> mCurrentState still live in the cluster. 

No, it lives in the application.  The cluster is in `src/app/clusters/media-playback-server/media-playback-server.{h,cpp}`.

> I assume the new function MediaPlaybackManager::HandleSetCurrentState(...) already fulfill this?

This is part of the _application_, not part of the _cluster_.  So every application would have to reimplement this part.

Maybe that's acceptable, but it seems a little odd to me to be pushing this work off on applications when we initially could not get it right in even our own example application.

Reply:
> So is what you do with Attributes::Whatever::Set, by default, unless it's either marked as NVM in the ZAP config or marked external and the external attribute handler does persistence.

Yeah. I plan to modify all Chef samples to use RAM instead of NVM as much as possible in a batch. But I can modify the BasicVideoPlayer first

> They might or might not be, depending on the ZAP configuration.

You mean defining it as external in ZAP. But I mean implement the `AttributeAccessInterface` (or `AAI`) in `src/app/clusters` . e.g. [WriteSingleClusterData](https://github.com/project-chip/connectedhomeip/blob/dd8b002e7e65d3f6b1f1a95ef8a740607202ea38/src/app/util/ember-compatibility-functions.cpp#L723-L733) which is not depending on the ZAP file.  My understanding is this is delegating read/write logics to application (if application registering a AAI implementation to it), and the application has to handle storage and report data itself.  It might be a little similar to `setting Attribute access as external`.

```
    if (auto * attrOverride = AttributeAccessInterfaceRegistry::Instance().Get(aPath.mEndpointId, aPath.mClusterId))
    {
        AttributeValueDecoder valueDecoder(aReader, aSubjectDescriptor);
        ReturnErrorOnFailure(attrOverride->Write(aPath, valueDecoder));


        if (valueDecoder.TriedDecode())
        {
            MatterReportingAttributeChangeCallback(aPath);
            return apWriteHandler->AddStatus(aPath, Protocols::InteractionModel::Status::Success);
        }
    }
```

>>  mCurrentState still live in the cluster.

> No, it lives in the application. The cluster is in src/app/clusters/media-playback-server/media-playback-server.{h,cpp}.

Got it.

> Maybe that's acceptable, but it seems a little odd to me to be pushing this work off on applications when we initially could not get it right in even our own example application.

This is how the current `AAI` is designed, isn't it?

Reply:
@bzbarsky-apple I've modified the ZAP to put some attribute to be in NVM. [ef5a9b](https://github.com/project-chip/connectedhomeip/pull/35122/commits/ef5a9bff1c1b464319fc6e2b1a7e952f65583a74). Would you be able to  take a look ?

Reply:
@bzbarsky-apple @tcarmelveilleux mPlaybackSpeed and mCurrentState are removed. PTAL